internal/mithril-metric/src/metric.rs

```diff
@@ -0,0 +1,139 @@
+use prometheus::{core::Collector, Counter, Gauge, Opts};
+use slog::{debug, Logger};
+
+use mithril_common::StdResult;
+
+/// Type alias for a metric name.
+pub type MetricName = str;
+
+/// Type alias for a counter value.
+pub type CounterValue = u32;
+
+/// Metric collector
+pub trait MetricCollector {
+    /// Metric name
+    fn name(&self) -> String;
+
+    /// Wrapped prometheus collector
+    fn collector(&self) -> Box<dyn Collector>;
+}
+
+/// Metric counter
+pub struct MetricCounter {
+    name: String,
+    logger: Logger,
+    counter: Box<Counter>,
+}
+
+impl MetricCounter {
+    pub fn new(logger: Logger, name: &str, help: &str) -> StdResult<Self> {
+        let counter = MetricCounter::create_metric_counter(name, help)?;
+        Ok(Self {
+            logger,
+            name: name.to_string(),
+            counter: Box::new(counter),
+        })
+    }
+
+    pub fn increment(&self) {
+        debug!(self.logger, "Incrementing '{}' counter", self.name);
+        self.counter.inc();
+    }
+
+    pub fn get(&self) -> CounterValue {
+        self.counter.get().round() as CounterValue
+    }
+
+    fn create_metric_counter(name: &MetricName, help: &str) -> StdResult<Counter> {
+        let counter_opts = Opts::new(name, help);
+        let counter = Counter::with_opts(counter_opts)?;
+
+        Ok(counter)
+    }
+}
+
+impl MetricCollector for MetricCounter {
+    fn collector(&self) -> Box<dyn Collector> {
+        self.counter.clone()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+/// Metric gauge
+pub struct MetricGauge {
+    name: String,
+    logger: Logger,
+    gauge: Box<Gauge>,
+}
+
+impl MetricGauge {
+    pub fn new(logger: Logger, name: &str, help: &str) -> StdResult<Self> {
+        let gauge = MetricGauge::create_metric_gauge(name, help)?;
+        Ok(Self {
+            logger,
+            name: name.to_string(),
+            gauge: Box::new(gauge),
+        })
+    }
+
+    pub fn record<T: Into<f64> + Copy>(&self, value: T) {
```

Comment:
I'm not sure that `Copy` is mandatory for this generic type but it can be fixed another time.

Reply:
Done